From 59bc4a7dbbaad06f7e4c9dd838b2af0d4d31fa3e Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Thu, 10 Apr 2025 23:25:43 +0530
Subject: [PATCH] mac80211: Add Mhz support

Add 320 MHz support in the channel def and handle center frequency
validation with compatible check.

Signed-off-by: Karthikeyan Periyasamy <quic_periyasa@quicinc.com>
Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 net/mac80211/mesh.c | 1 +
 net/wireless/chan.c | 5 +++++
 2 files changed, 6 insertions(+)

diff --git a/net/mac80211/mesh.c b/net/mac80211/mesh.c
index 1e95aea..a696380 100644
--- a/net/mac80211/mesh.c
+++ b/net/mac80211/mesh.c
@@ -1078,6 +1078,7 @@ ieee80211_mesh_build_beacon(struct ieee80211_if_mesh *ifmsh)
 		case NL80211_CHAN_WIDTH_80:
 		case NL80211_CHAN_WIDTH_80P80:
 		case NL80211_CHAN_WIDTH_160:
+		case NL80211_CHAN_WIDTH_320:
 			/* Channel Switch Wrapper + Wide Bandwidth CSA IE */
 			ie_len = 2 + 2 +
 				 sizeof(struct ieee80211_wide_bw_chansw_ie);
diff --git a/net/wireless/chan.c b/net/wireless/chan.c
index 1fbe653..328bb10 100644
--- a/net/wireless/chan.c
+++ b/net/wireless/chan.c
@@ -399,6 +399,11 @@ bool cfg80211_chandef_valid(const struct cfg80211_chan_def *chandef)
 		    chandef->center_freq2 - chandef->center_freq1 == 80)
 			return false;
 		break;
+	case NL80211_CHAN_WIDTH_160:
+	case NL80211_CHAN_WIDTH_320:
+		/* while center_freq2 is not mandatory, donot reject when
+		 * this is specified */
+		break;
 	default:
 		if (chandef->center_freq2)
 			return false;
-- 
2.34.1

